Euler discovered the remarkable quadratic formula:

n<sup>2</sup>+n+41
It turns out that the formula will produce 40 primes for the consecutive integer values 0≤n≤39. However, when n=40,402+40+41=40(40+1)+41 is divisible by 41, and certainly when n=41,412+41+41 is clearly divisible by 41.

The incredible formula n<sup>2</sup>−79n+1601 was discovered, which produces 80 primes for the consecutive values 0≤n≤79. The product of the coefficients, −79 and 1601, is −126479.

Considering quadratics of the form:

n<sup>2</sup>+an+b, where |a|<1000 and |b|≤1000

where |n| is the modulus/absolute value of n
e.g. |11|=11 and |−4|=4
Find the product of the coefficients, a and b, for the quadratic expression that produces the maximum number of primes for consecutive values of n, starting with n=0.

In [10]:
import math

def check_prime(num):
    prime = True
    for x in range(1,math.ceil(math.sqrt(num))+1):
        if num%x == 0 and x != 1:
            prime = False
            break
    return prime

In [11]:
prime_list=[1] 

for x in range(2,1000):
    if check_prime(x):
        prime_list.append(x)
        
print (prime_list)

[1, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]


## Observation 

### Obs 1
1, 41, 79 and 1601 are all primes  
a and b are probably primes too

### Obs 2
the first limit is 39, then 79, next maybe 119?
n = 199 = prime?

### Obs 3
When n = 40 or 80, the new number is divisble by 41

So..
119^2 +- 119*a = b   
120^2 +- 120*a -b % 41 = 0

In [15]:
import itertools

print([x for x in itertools.combinations(prime_list, 2)])

[(1, 3), (1, 5), (1, 7), (1, 11), (1, 13), (1, 17), (1, 19), (1, 23), (1, 29), (1, 31), (1, 37), (1, 41), (1, 43), (1, 47), (1, 53), (1, 59), (1, 61), (1, 67), (1, 71), (1, 73), (1, 79), (1, 83), (1, 89), (1, 97), (1, 101), (1, 103), (1, 107), (1, 109), (1, 113), (1, 127), (1, 131), (1, 137), (1, 139), (1, 149), (1, 151), (1, 157), (1, 163), (1, 167), (1, 173), (1, 179), (1, 181), (1, 191), (1, 193), (1, 197), (1, 199), (1, 211), (1, 223), (1, 227), (1, 229), (1, 233), (1, 239), (1, 241), (1, 251), (1, 257), (1, 263), (1, 269), (1, 271), (1, 277), (1, 281), (1, 283), (1, 293), (1, 307), (1, 311), (1, 313), (1, 317), (1, 331), (1, 337), (1, 347), (1, 349), (1, 353), (1, 359), (1, 367), (1, 373), (1, 379), (1, 383), (1, 389), (1, 397), (1, 401), (1, 409), (1, 419), (1, 421), (1, 431), (1, 433), (1, 439), (1, 443), (1, 449), (1, 457), (1, 461), (1, 463), (1, 467), (1, 479), (1, 487), (1, 491), (1, 499), (1, 503), (1, 509), (1, 521), (1, 523), (1, 541), (1, 547), (1, 557), (1, 563), (1, 56

My hypothesis didnt work, lets brute force this

In [108]:
def check_hypothesis(prime_list):
    best_n = 0
    best_tup = (1,2)
    for a in range(-1000,1001):
        for b in range(-1000, 1001):
            n = 0
            pos = n**2 + a*n + b
            
            if pos <= 1:
                continue
            
            while check_prime(pos):
                n += 1
                pos = math.pow(n,2) + n*a + b
                if pos <= 1:
                    break

            if n > best_n:
                best_tup = (a,b)
                best_n = n
    
    return best_n, best_tup

In [109]:
best_n, best_tup = check_hypothesis(prime_list)

In [105]:
best_n

71

In [106]:
best_tup

(-61, 971)

# FINAL ANSWER: -59231